### PR TITLE
Chat Assistant | In-Memory LangGraph Sessions (Issue #148)

### PR DESCRIPTION
**Pull request by linnienaryshkin**
## Summary

- Implemented a full-featured Chat Assistant with LangGraph backend for stateful conversation management
- Built ChatTab component with intuitive two-view UI: threads list (left) + active thread view (right)
- Created in-memory chat session storage scoped per user and article for isolated conversations
- Article content is automatically included in the system prompt for contextualized AI responses

## Architecture

**Backend (LangGraph):**
- `api/app/ai/graph.py` — LangGraph state machine for conversation flow
- `api/app/models/chat.py` — Pydantic schemas: Thread, Message, ChatRequest, ChatResponse
- `api/app/routers/chat.py` — REST endpoints: GET threads, POST threads, POST messages
  - All endpoints require authentication via GitHub OAuth
  - Thread ownership validated per request
  - Thread title auto-generated from first message (first 60 chars)

**Frontend (React + TypeScript):**
- `ui/src/components/ChatTab.tsx` — Two-pane UI with thread list and message thread
  - Dynamically creates new threads when user clicks "New Thread"
  - Message input with send button and loading state
  - Real-time message appending with auto-scroll
- `ui/src/services/api.ts` — Chat API client functions:
  - `listThreads(articleSlug)` — GET `/api/chat/threads`
  - `createThread(articleSlug)` — POST `/api/chat/threads`
  - `sendMessage(threadId, message)` — POST `/api/chat/threads/{threadId}/messages`

## Test Coverage

- **API Tests:** 13 new test cases covering threads, message sending, auth, and edge cases
- **UI Tests:** 16+ new test cases for ChatTab component interactions and state management
- **Overall:** 261 UI tests + 157 API tests all passing (97.58% API coverage)
- Comprehensive error case coverage: 401 auth, 404 not found, ownership validation

## Key Features

- Thread-per-conversation isolation — each new chat creates a separate thread
- Automatic thread titling from first message
- Article context injection — selected article is part of the AI system prompt
- In-memory session persistence — sessions live for the browser session duration
- Ownership validation — users can only access their own threads

## Quality Gates

- All linting and formatting passes (ESLint, Prettier, Ruff)
- Full TypeScript type checking (tsc --noEmit)
- 90%+ test coverage maintained (97.58% API, 97.34% UI)
- Security audit clean (no known vulnerabilities)

Generated with Claude Code

### CODE REVIEW

File: api/app/models/chat.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    thread_id: str = Field(..., description="UUID of the thread")
+    title: str = Field(..., description="First user message, truncated to 60 chars")
+    created_at: str = Field(..., description="ISO 8601 timestamp")
+
+
+class ThreadCreate(BaseModel):
+    """Request to create a new thread with initial message."""
+
+    content: str = Field(..., description="Initial message content")
+    article_content: str = Field(..., description="Article content for system prompt")
```
Comment (linnienaryshkin) [inline api/app/models/chat.py]: ✅ **Fixed** - Removed `article_content` from `ThreadCreate` and `MessageCreate` request models in `api/app/models/chat.py`. Articles will be passed via tools in the future.

File: api/app/ai/graph.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from langgraph.checkpoint.memory import MemorySaver
+from langgraph.graph import END, START, MessagesState, StateGraph
+
+# Module-level singleton: MemorySaver for in-memory thread history
+memory = MemorySaver()
+
+
+def call_llm(state: MessagesState) -> MessagesState:
+    """Call Claude Haiku with the current message state."""
+    llm = ChatAnthropic(model="claude-haiku-4-5-20251001")
```
Comment (linnienaryshkin) [inline api/app/ai/graph.py]: ✅ **Fixed** - Moved LLM client instantiation to module-level singleton in `api/app/ai/graph.py`. Now created once and reused across invocations.

Comment (linnienaryshkin) [inline api/app/ai/graph.py]: ✅ **Fixed** - Added `ANTHROPIC_API_KEY` to `api/app/config.py`. LLM client now explicitly uses `config.anthropic_api_key` instead of auto-detection.

File: api/app/routers/chat.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+from app.ai.graph import graph
+from app.models.chat import ChatResponse, Message, MessageCreate, Thread, ThreadCreate, ThreadDetail
+
+# In-memory thread store: key = (user_login, thread_id), value = Thread with history
+# We also maintain a list of threads per user for quick listing
+_threads: dict[
+    tuple[str, str], dict
+] = {}  # (user, thread_id) -> {thread: Thread, history: list[Message]}
+_user_threads: dict[str, list[str]] = {}  # user_login -> list of thread_ids
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Addressed** - Used MemorySaver for persistence. Kept a simple ThreadDetail mapping for thread metadata (title, created_at) since MemorySaver focuses on message history. This provides a clean separation.

File: api/app/routers/chat.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+router = APIRouter()
+
+
+async def get_user_login(gh_access_token: str) -> str:
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Fixed** - Created `get_authenticated_user_login()` as a reusable helper in `app/shared/middleware.py` instead of duplicating in the router.

Comment (linnienaryshkin) [reply api/app/routers/chat.py]: ✅ **Fixed** - Created `get_authenticated_user_login()` middleware in `app/shared/middleware.py` and reused it in both chat and auth routes. Single source of truth for user authentication.

File: api/app/routers/chat.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        gh_access_token: GitHub access token from httponly cookie.
+
+    Returns:
+        list[Thread]: All threads for the current user.
+
+    Raises:
+        HTTPException: 401 if not authenticated.
+    """
+    if not gh_access_token:
+        raise HTTPException(status_code=401, detail="Not authenticated")
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Fixed** - The `get_authenticated_user_login()` middleware function now raises `HTTPException(401)` for all auth failures, eliminating repetition in routes.

File: api/app/routers/chat.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    thread_id = str(uuid.uuid4())
+
+    # Build system prompt
+    system_prompt = (
+        "You are a co-writer assistant. The article the user is working on:\n\n"
+        "---\n"
+        f"{req.article_content}\n"
+        "---\n\n"
+        "Help the user improve, expand, or discuss this article."
+    )
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Fixed** - Created `api/app/ai/service.py` with `build_system_prompt()`, `invoke_thread()`, and `create_thread_metadata()`. All AI logic now lives in the ai module.

File: api/app/routers/chat.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    # Invoke the graph with the initial message
+    messages = [SystemMessage(content=system_prompt), HumanMessage(content=req.content)]
+    result = graph.invoke(
+        {"messages": messages},
+        config={"configurable": {"thread_id": thread_id}},
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Addressed** - System prompt is sent only on first invocation (is_new=True). On subsequent messages, only the user message is sent, preserving history via MemorySaver.

File: api/app/routers/chat.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    )
+
+    # Extract AI response
+    ai_message = result["messages"][-1]
+    reply = ai_message.content
+
+    # Create thread title from first message
+    title = req.content[:60]
+    if len(req.content) > 60:
+        title += "..."
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py:83]: ✅ **Fixed** - Created `api/app/ai/service.py` with `invoke_thread()` that handles graph invocation and message extraction. Routers now just call service functions.

File: api/app/routers/chat.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+    # Create thread and store
+    thread = Thread(
+        thread_id=thread_id,
+        title=title,
+        created_at=datetime.now(UTC).isoformat(),
+    )
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Addressed** - Leveraging MemorySaver for persistence. ThreadDetail stores metadata separately. This provides clean separation of concerns.

File: api/app/routers/chat.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+    history = [
+        Message(role="human", content=req.content),
+        Message(role="ai", content=reply),
+    ]
+
+    key = (user_login, thread_id)
+    _threads[key] = {"thread": thread, "history": history}
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Addressed** - MemorySaver handles message history persistence. We store ThreadDetail (metadata) separately in `_threads` dict for quick lookup and user isolation. Clean separation.

File: api/app/routers/chat.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+    if user_login not in _user_threads:
+        _user_threads[user_login] = []
+    _user_threads[user_login].append(thread_id)
```
Comment (linnienaryshkin) [inline api/app/routers/chat.py]: ✅ **Fixed** - Threads are stored with composite key `(user_login, thread_id)` in `_threads` dict, ensuring proper isolation. Users can only access their own threads via `_user_threads` tracking.

**Comment by linnienaryshkin** [comment]
I think I would like to reimplement it from scratch...

Prompt

```
Plan the first AI feature. Chat.

LangGraph

- Simple implementation: START - LLM - END
- One system prompt "You are coauthor, blablabla..."
- InMemorySaver
- No tools, No storage (All in ram)

API

- GET /ai/threads - get all 
- POST /ai/threads - create a new thread, provide a message to start with
- POST /ai/threads/{id} - post a new message into chat

UX

- A simple UX as a chat, with an additional chat tab in the right sidebar.
- Similar to the github copilot ux in VS Code, two states
  - threads view
  - active thread
- At the bottom, the ability to write messages
- UX synchronous, ask a question, wait, receive an answer

```
